File: backend/models/habit_streak.py
"""
Habit Streak database model
"""
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey
from sqlalchemy.dialects.mysql import CHAR
from sqlalchemy.orm import relationship
from database.database import Base


class HabitStreak(Base):
    __tablename__ = "habit_streaks"

    id = Column(CHAR(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(CHAR(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    habit_name = Column(String(100), nullable=False)
    current_streak = Column(Integer, default=0)
    best_streak = Column(Integer, default=0)
    last_completed = Column(DateTime, nullable=True)
    target_frequency = Column(String(50), nullable=True)  # daily, weekly
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="habit_streaks")

File: lib/models/habit_streak.dart
"""
HabitStreak model mirroring backend schema
"""
class HabitStreak {
  final String id;
  final String userId;
  final String habitName;
  final int currentStreak;
  final int bestStreak;
  final DateTime? lastCompleted;
  final String? targetFrequency;
  final DateTime createdAt;

  const HabitStreak({
    required this.id,
    required this.userId,
    required this.habitName,
    this.currentStreak = 0,
    this.bestStreak = 0,
    this.lastCompleted,
    this.targetFrequency,
    required this.createdAt,
  });

  factory HabitStreak.fromJson(Map<String, dynamic> j) => HabitStreak(
    id: j['id'] as String,
    userId: j['user_id'] as String,
    habitName: j['habit_name'] as String,
    currentStreak: j['current_streak'] as int? ?? 0,
    bestStreak: j['best_streak'] as int? ?? 0,
    lastCompleted: j['last_completed'] != null
        ? DateTime.parse(j['last_completed'] as String)
        : null,
    targetFrequency: j['target_frequency'] as String?,
    createdAt: DateTime.parse(j['created_at'] as String),
  );

  Map<String, dynamic> toJson() => {
    'habit_name': habitName,
    'target_frequency': targetFrequency,
  };
}
